lib/redirect-agent.js

```diff
@@ -0,0 +1,171 @@
+    this[kAgent].on('disconnect', (...args) => this.emit(...args))
+  }
+
+  dispatch (origin, { maxRedirections = this[kMaxRedirections], ...opts }, handler) {
+    if (maxRedirections != null && (maxRedirections < 0 || !Number.isInteger(maxRedirections))) {
+      throw new InvalidArgumentError('maxRedirections must be a positive number')
+    }
+
+    if (!maxRedirections) {
+      return this[kAgent].dispatch(origin, opts, handler)
+    }
+
+    if (util.isStream(opts.body) && util.bodyLength(opts.body) === 0) {
+      opts.body
+        .on('data', /* istanbul ignore next */ function () {
+          /* istanbul ignore next */
+          assert(false)
+        })
+        .on('error', function (err) {
+          // TODO (fix): What to do?
+          throw err
+        })
+        .on('end', function () {
+          util.destroy(this)
+        })
+      opts.body = null
+    }
+
+    if (util.isStream(opts.body)) {
+      return this[kAgent].dispatch(origin, opts, handler)
+    }
+
+    opts.maxRedirections = maxRedirections - 1
+
+    this[kAgent].dispatch(origin, opts, {
+      agent: this,
+      location: null,
+      opts,
+      origin,
+      handler,
+
+      onConnect (abort) {
+        this.onConnect(abort)
+      },
+
+      onUpgrade (statusCode, headers, socket) {
+        this.onUpgrade(statusCode, headers, socket)
+      },
+
+      onError (error) {
+        this.onError(error)
+      },
+
+      onHeaders (statusCode, headers, resume) {
+        if ([300, 301, 302, 303, 307, 308].indexOf(statusCode) === -1) {
+          return this.handler.onHeaders(statusCode, headers, resume)
+        }
+
+        this.location = parseLocation(headers)
+
+        if (!this.location) {
+          return this.handler.onHeaders(statusCode, headers, resume)
+        }
+
+        this.location = new URL(this.location, this.origin)
+
+        if (!this.opts.history) {
+          opts.history = new Set()
+        } else if (this.opts.history.has(this.location.href)) {
+          return this.handler.onHeaders(statusCode, headers, resume)
+        }
+
+        this.opts.history.add(this.location.href)
+
+        // Remove headers referring to the original URL.
+        // By default it is Host only, unless it's a 303 (see below), which removes also all Content-* headers.
+        // https://tools.ietf.org/html/rfc7231#section-6.4
+        this.opts.headers = cleanRequestHeaders(this.opts.headers, statusCode === 303)
+
+        // https://tools.ietf.org/html/rfc7231#section-6.4.4
+        // In case of HTTP 303, always replace method to be either HEAD or GET
+        if (statusCode === 303) {
+          this.opts.method = 'GET'
+          this.opts.body = null
+        }
+      },
+
+      onData (chunk) {
+        if (!this.location) {
+          return this.handler.onData(chunk)
+        }
+      },
+
+      onComplete (trailers) {
+        if (this.location) {
+          this.dispatch(this.location, this.opts, this.handler)
```

Comment:
Little note here.
If you trigger dispatch from here, it means you will wait for the body (which, even not common, are legit) to be transferred before following the redirect.

That's why in #600 I dispatched from `onHeaders`. Especially when going cross-origin, we don't care about the original response body, isn't it?

Reply:
Unfortunately we need to receive the entire body or pipelining won't work.

Reply:
Right point. 

I guess we need two approaches:

1. Same origin -> dispatch from onComplete
2. Cross origin -> dispatch from onHeaders

This way the client will experience the best performance. Am I wrong?

Reply:
Yes, I think we are both right actually. We must read the body, BUT you are right that we could actually dispatch the follow immediately in `onHeaders`.

Reply:
The only problem with dispatching in onHeaders is that I'm unsure how to handle any error to might occur while reading the original request.

Reply:
Yeah, I come to the same conclusion XD

Reply:
I'm a little unsure whether the added complexity is worth the performance gain which I'm not sure whether are relevant in practice (though I have no numbers to back my scepticism).

Reply:
Usually a redirect request will receive the EOF in the same TCP packet so `onComplete` is called immediatly.

Reply:
To be honest I have no idea as well. 

It's unlikely that servers will send big bodies in case of 3xx. But I don't trust any server XD

Reply:
You ok with leaving as is?

Reply:
If body is empty yes.

But I do also see you're point about error handling.

Well, then maybe just leave a comment note that this can technically be done so we can revisit in the future.

Reply:
Sorry, I replied before reading your last question.

Yes, I'm fine with leaving like it. Maybe, as I said above, leave a note about this possible cross-origin optimization.